File: HouseRob.java
//Time complexity O(n)
//space complexity O(1)

class HouseRob {

    //*****Exhaustive approach */
    // public int rob(int[] nums) {
    //     int  re = helper(nums,0);
    //     return re;
    // }
    // public static int helper(int[] nums, int idx){
    //     if(idx>=nums.length) return 0;
    //     //not choose
    //     int case1 = helper(nums, idx+1);
    //     //choose
    //     int case2 = nums[idx] + helper(nums, idx+2);
    //     return Math.max(case1, case2);
    // }

    
//*****Optimised solution */
    public int rob(int[] nums) {
        int n = nums.length;
        int prev = 0;
        int curr = nums[0];
        for (int i = 1; i < n; i++) {
            int temp = curr;
            curr = Math.max(curr, nums[i] + prev);
            prev = temp;
        }
        return curr;
    }

    public static void main(String[] args) {
        HouseRob obj = new HouseRob();
        int[] houses = {1, 2, 3, 1};
        System.out.println(obj.rob(houses));
    }
}

File: house_rob.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        if len(nums)==1:
            return nums[0]
        if not nums:
            return 0
        dp = nums
        dp[0]=nums[0]
        dp[1] = max(nums[0],nums[1])
        for i in range(2,len(nums)):
            dp[i] = max(nums[i-1],nums[i]+dp[i-2])
        
        print(dp)
        return dp[-1]
